File: scripts/process-photo.js
#!/usr/bin/env node
/**
 * Photo processor — duotone via contrast S-curve.
 * No background removal needed: mix-blend-mode:screen in CSS
 * makes pure-black pixels transparent on the dark site background.
 *
 * Usage:  node scripts/process-photo.js
 */

const path = require('path');
const fs   = require('fs');

const BASE     = path.resolve(__dirname, '..');
const IMG      = path.join(BASE, 'img');
const SOURCE   = path.join(IMG, 'founder-original.jpg');
const DUOTONE  = path.join(IMG, 'founder-duotone.png');
const DUOTONE2 = path.join(IMG, 'founder-duotone@2x.png');
const MAX_EDGE = 1200;

/* Duotone palette — pure black shadows so screen blend erases them */
const SHADOW    = { r: 0,   g: 0,   b: 0   };  // #000000
const HIGHLIGHT = { r: 201, g: 243, b: 29  };  // #C9F31D

module.paths.unshift(path.join(BASE, 'node_modules'));

/* S-curve contrast: darkens shadows, boosts highlights */
function sCurve(t, strength = 2.2) {
  if (t <= 0) return 0;
  if (t >= 1) return 1;
  if (t < 0.5) return 0.5 * Math.pow(2 * t, strength);
  return 1 - 0.5 * Math.pow(2 * (1 - t), strength);
}

async function main() {
  console.log('\n── Web Design Mindset · Photo Processor ──\n');

  if (!fs.existsSync(SOURCE)) {
    console.error('ERROR: img/founder-original.jpg not found.');
    process.exit(1);
  }

  const { Jimp } = require('jimp');

  console.log('Loading  founder-original.jpg…');
  const srcBuf = fs.readFileSync(SOURCE);
  const img    = await Jimp.fromBuffer(srcBuf);
  const { width, height } = img.bitmap;
  console.log(`  ${width}×${height}px`);

  console.log('\nApplying contrast-enhanced duotone…');

  const data = img.bitmap.data;

  for (let i = 0; i < width * height; i++) {
    const base = i * 4;
    const r = data[base];
    const g = data[base + 1];
    const b = data[base + 2];

    /* Luminance-weighted grayscale */
    const gray = (0.2126 * r + 0.7152 * g + 0.0722 * b) / 255;

    /* S-curve contrast: crush dark values to black, pull highlights up */
    const t = sCurve(gray, 2.2);

    /* Map to duotone gradient */
    data[base]     = Math.round(SHADOW.r + t * (HIGHLIGHT.r - SHADOW.r));
    data[base + 1] = Math.round(SHADOW.g + t * (HIGHLIGHT.g - SHADOW.g));
    data[base + 2] = Math.round(SHADOW.b + t * (HIGHLIGHT.b - SHADOW.b));
    data[base + 3] = 255; /* fully opaque — CSS blend mode handles transparency */
  }

  /* Save full-size duotone */
  const duoBuf = await img.getBuffer('image/png');
  fs.writeFileSync(DUOTONE, duoBuf);
  console.log('  ✓ Saved  founder-duotone.png');

  /* Web-optimised copy ≤ 1200px */
  const longest = Math.max(width, height);
  const ratio   = longest > MAX_EDGE ? MAX_EDGE / longest : 1;
  const ow = Math.round(width  * ratio);
  const oh = Math.round(height * ratio);

  const img2 = await Jimp.fromBuffer(duoBuf);
  img2.resize({ w: ow, h: oh });
  const optBuf = await img2.getBuffer('image/png');
  fs.writeFileSync(DUOTONE2, optBuf);
  console.log(`  ✓ Saved  founder-duotone@2x.png  (${ow}×${oh})`);

  console.log('\n── Done! ──');
  console.log('CSS mix-blend-mode:screen will make the black areas transparent');
  console.log('on the dark site background — no background removal needed.\n');
}

main().catch(err => { console.error(err); process.exit(1); });

File: scripts/process-photo.py
#!/usr/bin/env python3
"""
Photo processing for Web Design Mindset.
Place your portrait at img/founder-original.jpg, then run:
    python scripts/process-photo.py
"""

import subprocess
import sys
import os
from pathlib import Path

BASE_DIR = Path(__file__).parent.parent
IMG_DIR  = BASE_DIR / "img"
SOURCE   = IMG_DIR / "founder-original.jpg"
CUTOUT   = IMG_DIR / "founder-cutout.png"
DUOTONE  = IMG_DIR / "founder-duotone.png"
DUOTONE2 = IMG_DIR / "founder-duotone@2x.png"

SHADOW_COLOR    = (10, 10, 10)       # #0A0A0A
HIGHLIGHT_COLOR = (201, 243, 29)     # #C9F31D
MAX_LONG_EDGE   = 1200


def pip_install(package: str) -> None:
    try:
        __import__(package.split("[")[0].replace("-", "_"))
    except ImportError:
        print(f"  Installing {package}…")
        subprocess.check_call([sys.executable, "-m", "pip", "install", package, "-q"])


def apply_duotone(rgba_image):
    from PIL import Image, ImageOps
    r, g, b, a = rgba_image.split()
    gray = ImageOps.grayscale(rgba_image.convert("RGB"))

    lut_r, lut_g, lut_b = [], [], []
    for i in range(256):
        t = i / 255.0
        lut_r.append(int(SHADOW_COLOR[0] + t * (HIGHLIGHT_COLOR[0] - SHADOW_COLOR[0])))
        lut_g.append(int(SHADOW_COLOR[1] + t * (HIGHLIGHT_COLOR[1] - SHADOW_COLOR[1])))
        lut_b.append(int(SHADOW_COLOR[2] + t * (HIGHLIGHT_COLOR[2] - SHADOW_COLOR[2])))

    from PIL import Image
    duo = Image.merge("RGB", (
        gray.point(lut_r),
        gray.point(lut_g),
        gray.point(lut_b),
    )).convert("RGBA")
    duo.putalpha(a)
    return duo


def resize_web(image, max_edge=MAX_LONG_EDGE):
    from PIL import Image
    w, h = image.size
    if max(w, h) <= max_edge:
        return image
    ratio = max_edge / max(w, h)
    return image.resize((int(w * ratio), int(h * ratio)), Image.LANCZOS)


def remove_bg_pillow_fallback(image):
    """Flood-fill from all four edges to erase near-uniform backgrounds."""
    from PIL import Image
    from collections import deque

    rgba  = image.convert("RGBA")
    data  = rgba.load()
    W, H  = rgba.size

    corners = [data[0, 0][:3], data[W-1, 0][:3], data[0, H-1][:3], data[W-1, H-1][:3]]
    bg = tuple(int(sum(c[i] for c in corners) / 4) for i in range(3))
    tol = 45

    visited = set()
    q = deque()
    for x in range(W):
        q.append((x, 0)); q.append((x, H - 1))
    for y in range(H):
        q.append((0, y)); q.append((W - 1, y))

    while q:
        x, y = q.popleft()
        if (x, y) in visited or not (0 <= x < W and 0 <= y < H):
            continue
        r, g, b, a = data[x, y]
        if all(abs(int(c) - int(bg[i])) <= tol for i, c in enumerate((r, g, b))):
            data[x, y] = (r, g, b, 0)
            visited.add((x, y))
            q.extend([(x+1, y), (x-1, y), (x, y+1), (x, y-1)])

    return rgba


def main():
    print("\n── Web Design Mindset · Photo Processor ──\n")

    if not SOURCE.exists():
        print(f"ERROR: {SOURCE} not found.\n")
        print("Drop your portrait at  img/founder-original.jpg  then re-run this script.")
        sys.exit(1)

    print("Installing / verifying dependencies…")
    pip_install("Pillow")
    pip_install("onnxruntime")
    pip_install("rembg")

    from PIL import Image
    import io

    print(f"\nLoading  {SOURCE.name}…")
    original = Image.open(SOURCE)

    try:
        from rembg import remove
        print("Removing background with rembg…")
        with open(SOURCE, "rb") as f:
            raw = f.read()
        cutout = Image.open(io.BytesIO(remove(raw))).convert("RGBA")
        print("  ✓ rembg succeeded")
    except Exception as exc:
        print(f"  rembg failed ({exc}) — using Pillow flood-fill fallback…")
        cutout = remove_bg_pillow_fallback(original)
        print("  ✓ Fallback complete")

    cutout.save(CUTOUT, "PNG")
    print(f"  ✓ Saved  {CUTOUT.name}")

    print("\nApplying duotone (#0A0A0A → #C9F31D)…")
    duotone = apply_duotone(cutout)

    duotone.save(DUOTONE, "PNG")
    print(f"  ✓ Saved  {DUOTONE.name}")

    optimised = resize_web(duotone)
    optimised.save(DUOTONE2, "PNG", optimize=True)
    print(f"  ✓ Saved  {DUOTONE2.name}  ({optimised.width}×{optimised.height})")

    print("\n── Done! Update img src in HTML from placehold.co to img/founder-duotone.png ──\n")


if __name__ == "__main__":
    main()
